implementations/rust/ockam/ockam/src/entity/profile/profile_change_type/create_key.rs

```diff
@@ -0,0 +1,100 @@
+use crate::entity::profile::event_identifier::EventIdentifier;
+use crate::entity::profile::profile_change::{ProfileChange, ProfileKeyPurpose, ProfileKeyType};
+use crate::entity::profile::profile_change_type::ProfileChangeType;
+use crate::entity::profile::proof::{ProfileChangeProof, Signature, SignatureType};
+use crate::entity::{Changes, Profile, ProfileChangeEvent, ProfileEventAttributes};
+use crate::OckamError;
+use ockam_vault_core::types::{SecretAttributes, SecretPersistence, SecretType};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CreateKeyChange {
+    label: String,
+    key_type: ProfileKeyType,
+    key_purpose: ProfileKeyPurpose,
+    public_key: Vec<u8>,
+}
+
+impl CreateKeyChange {
+    pub fn new(
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        public_key: Vec<u8>,
+    ) -> Self {
+        CreateKeyChange {
+            label,
+            key_type,
+            key_purpose,
+            public_key,
+        }
+    }
+}
+
+impl CreateKeyChange {
+    pub fn label(&self) -> &str {
+        &self.label
+    }
+    pub fn key_type(&self) -> ProfileKeyType {
+        self.key_type
+    }
+    pub fn key_purpose(&self) -> ProfileKeyPurpose {
+        self.key_purpose
+    }
+    pub fn public_key(&self) -> &[u8] {
+        &self.public_key
+    }
+}
+
+impl Profile {
+    pub(crate) fn create_key_event(
+        &mut self,
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+    ) -> Result<ProfileChangeEvent, ockam_core::Error> {
+        let attributes = attributes.unwrap_or(ProfileEventAttributes::new());
+
+        // Creating key after it was revoked is forbidden
+        if self
+            .find_last_key_event(&label, key_type, key_purpose)
+            .is_ok()
+        {
+            return Err(OckamError::InvalidInternalState.into());
+        }
+
+        let prev_id = self.get_last_event_id()?;
+
+        let mut v = self.vault.lock().unwrap();
+
+        // TODO: Should be customisable
+        let secret_attributes = SecretAttributes {
+            stype: SecretType::Curve25519,
+            persistence: SecretPersistence::Persistent,
+            length: 0,
```

Comment:
length could be derivable from key type.

Reply:
for private keys yes, for buffers - no. that's why we decided to add length as separate field during last vault rework

Reply:
In truth, when do we ever use a buffer? Buffers are just like an opaque type that we store with keys. I'm not sure it makes sense to put them there. I think we should move this to the knox layer.

Reply:
chain key is a good example of secret buffer